File: bintree_java/src/binarytree/traversal/BaseTraversal.java
package binarytree.traversal;


import binarytree.BinNode;
import java.util.List;



public abstract class BaseTraversal< TKey extends Comparable<? super TKey>,
                                     TNode extends BinNode<TKey, TNode>
                                   >
{
    public abstract List<TKey> traverse(TNode root, OrderTraversal order);
}

File: bintree_csharp/binarytree/traversal/BaseTraversal.cs
using System;
using System.Collections.Generic;


namespace my.binarytree.traversal
{
    abstract class BaseTraversal<TKey, TNode>
        where TKey : IComparable where TNode : BinNode<TKey, TNode>, new()
    {
        public abstract List<TKey> Traverse(TNode root, OrderTraversal order);
    }
}
